presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHadoopAuthentication.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.authentication;
+
+import com.facebook.presto.hive.HiveErrorCode;
+import com.facebook.presto.spi.PrestoException;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.authentication.util.KerberosName;
+
+import javax.security.auth.Subject;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class KerberosHadoopAuthentication
+        implements HadoopAuthentication
+{
+    static {
+        // In order to enable KERBEROS authentication method for HDFS
+        // UserGroupInformation.authenticationMethod static field must be set to KERBEROS
+        // It is further used in many places in DfsClient
+        Configuration configuration = new Configuration(false);
+        configuration.set("hadoop.security.authentication", "kerberos");
+        UserGroupInformation.setConfiguration(configuration);
+
+        // KerberosName#rules static field must be initialized
+        // It is used in KerberosName#getShortName which is used in User constructor invoked by UserGroupInformation#getUGIFromSubject
+        KerberosName.setRules("DEFAULT");
+    }
+
+    private final KerberosAuthentication kerberosAuthentication;
+
+    public KerberosHadoopAuthentication(KerberosAuthentication kerberosAuthentication)
+    {
+        this.kerberosAuthentication = requireNonNull(kerberosAuthentication, "kerberosAuthentication is null");
+    }
+
+    @Override
+    public UserGroupInformation getUserGroupInformation()
+    {
+        Subject subject = kerberosAuthentication.getSubject();
+        return createUserGroupInformationForSubject(subject);
+    }
+
+    private UserGroupInformation createUserGroupInformationForSubject(Subject subject)
+    {
+        checkState(KerberosName.hasRulesBeenSet(), "UserGroupInformation has not been initialized with configuration, or the rules was reset");
+        Method createUGIMethod = getCreateUGIMethod();
```

Comment:
Same here.  We can add a class to the UGI package and access the protected constructor directly.